File: src/main/java/algo/slidingWindow/LongestSubstringWithoutRepeatingCharacters_3.java
package algo.slidingWindow;

import java.util.Arrays;

public class LongestSubstringWithoutRepeatingCharacters_3 {
  public static int lengthOfLongestSubstring(String s) {
    int i = 0, j = 0;
    int maxlen = 0;
    var freq = new int[256];
    Arrays.fill(freq, 0);

    while (j < s.length()) {
      var c = s.charAt(j);

      // move left window till there is no repeats
      if (freq[c] != 0) {
        while (freq[c] != 0) {
          freq[s.charAt(i)]--;
          i++;
        }
      }

      freq[c]++;
      maxlen = Math.max(maxlen, j - i + 1);
      j++;
    }

    return maxlen;
  }

  public static void main(String[] args) {
    System.out.println(lengthOfLongestSubstring("abcabcbb"));
    System.out.println(lengthOfLongestSubstring("bbbbb"));
    System.out.println(lengthOfLongestSubstring("pwwkew"));
  }
}

File: src/main/kotlin/algo/slidingWindow/LongestSubstringWithoutRepeatingCharacters_3.kt
package algo.slidingWindow

class KtLongestSubstringWithoutRepeatingCharacters_3 {
    fun lengthOfLongestSubstring(s: String): Int {
        val count: MutableSet<Char> = mutableSetOf()
        var left = 0; var right = 0
        var maxLen = 0
        while (right < s.length) {
            val ch = s[right]
            if (!count.add(ch)) {
                while (s[left] != ch) {
                    count.remove(s[left])
                    left++
                }
                left++
            } else {
                maxLen = maxLen.coerceAtLeast(right - left + 1)
            }
            right++
        }
        return maxLen
    }
}

fun main() {
    val substring = KtLongestSubstringWithoutRepeatingCharacters_3()
    println(substring.lengthOfLongestSubstring("abcabcbb"))
    println(substring.lengthOfLongestSubstring("bbbbb"))
    println(substring.lengthOfLongestSubstring("pwwkew"))
}